### PR TITLE
[Merged by Bors] - now both hare and tortoise can update state

### LINKED ISSUE #1805: Update global state with TXs once hare terminates
When the Hare protocol terminates on layer j, if the global state is updated to layer j-1, we update it according to the blocks in the Hare output set. If not, we wait until all (open) Hare consensus processes in previous layers terminate, then we update the state, after updating layer j-1.

Possibly, the logic was already applied here: https://github.com/spacemeshos/local-testnet/issues/10

### PR DESCRIPTION
**Pull request by antonlerner**
## Motivation
Closes #1805 

## Changes
added a unified method in mesh to update the state `updateStateWithLayer`
Now inputs can be received via 2 code paths, the first is `ValidateLayer` which runs tortoise validation and then update the state
Note: this may result in different states received when running a node and receiving inputs from hare and syncing a node, since potentially, if assumptions are broken , tortoise may contain more blocks than first agreed by hare protocol for a certain layer @barakshani , please re affirm this

## Test Plan
add UT to test receiving inputs from both tortoise and hare,
test state received when syncing from tortoise vs state received from hare

### CODE REVIEW

**Comment by antonlerner** [comment]
bors try

**Comment by antonlerner** [comment]
bors try

**Comment by antonlerner** [comment]
bors try

**Comment by barakshani** [comment]
Yes, if the assumptions are broken then Tortoise and Hare may not agree (this is true in general). Indeed, when miners only run the tortoise (e.g. a syncing miner) then they may have a different state than miners that builds their state on top of the Hare.

**Comment by antonlerner** [comment]
bors try

**Comment by antonlerner** [comment]
bors try

**Comment by antonlerner** [comment]
bors try

**Comment by antonlerner** [comment]
bors try

File: mesh/mesh.go [hunk 1]
```diff
@@ -236,19 +240,65 @@ func (m *Mesh) pushLayersToState(oldPbase types.LayerID, newPbase types.LayerID)
...
-		m.logStateRoot(layerId)
-		m.setLayerHash(l)
-		m.setLatestLayerInState(layerId)
+		validBlocks, _ := m.BlocksByValidity(l.Blocks())
+		m.updateStateWithLayer(layerId, types.NewExistingLayer(layerId, validBlocks))
 	}
 	m.persistLayerHash()
 }
 
+func (m *Mesh) applyState(layerId types.LayerID, l *types.Layer) {
```
Comment (noamnelke) [inline mesh/mesh.go]: `types.Layer` exposes `.Index()` which returns the layer ID - I think it would be cleaner to use that instead of essentially passing the layer ID twice.

File: mesh/mesh.go [hunk 2]
```diff
@@ -236,19 +240,65 @@ func (m *Mesh) pushLayersToState(oldPbase types.LayerID, newPbase types.LayerID)
 	for layerId := oldPbase; layerId < newPbase; layerId++ {
 		l, err := m.GetLayer(layerId)
 		if err != nil || l == nil {
-			// TODO: propagate/handle error
```
Comment (noamnelke) [inline mesh/mesh.go:270]: I don't agree with removing this TODO. If fetching the layer from the DB fails, even when changing the `break` to a `return` this will essentially be ignored and the code will move to the next layer. This is tech debt - it requires proper handling. To improve the situation, a `panic` would be better suited here, IMO. Otherwise, assuming you don't want to now do proper error handling - perhaps just return the TODO?

File: mesh/mesh.go [hunk 3]
```diff
@@ -236,19 +240,65 @@ func (m *Mesh) pushLayersToState(oldPbase types.LayerID, newPbase types.LayerID)
...
+}
+
+func (m *Mesh) HandleValidatedLayer(validatedLayer types.LayerID, layer []types.BlockID) {
+	blocks := []*types.Block{}
+
+	for _, blockId := range layer {
+		block, err := m.GetBlock(blockId)
+		if err != nil {
+			//todo: can this happen?
+			log.Panic("hare terminated with block that is not present in mesh")
```
Comment (noamnelke) [inline mesh/mesh.go]: Hare explicitly supports terminating with an unfamiliar block (it used to `panic` if this happened, but that was removed because it happens from time to time). If we encounter this we should try to fetch the missing block from our neighbors, if not here then somewhere else, but `panic`ing here is unacceptable until then, IMO.

File: mesh/meshdb.go [hunk 4]
```diff
@@ -469,23 +469,18 @@ func (m *MeshDB) addToAccountTxs(addr types.Address, accountTxs []*types.Transac
 	return nil
 }
 
-func (m *MeshDB) removeFromUnappliedTxs(accepted, rejected []*types.Transaction, layer types.LayerID) {
```
Comment (noamnelke) [inline mesh/meshdb.go:476]: I don't understand the changes to this file. When a block is added to the mesh, its transactions are added to the pending transactions list. After this change, if a transaction only appears in a contextually invalid block it will never be removed from the pending transactions list. Am I missing something?

File: mesh/mesh.go [hunk 5]
```diff
@@ -337,26 +387,17 @@ func (m *Mesh) getTxs(txIds []types.TransactionId, l *types.Layer) []*types.Tran
...
 	}
-	m.removeFromUnappliedTxs(validBlockTxs, invalidBlockTxs, l.Index())
-	for _, tx := range invalidBlockTxs {
-		err = m.blockBuilder.ValidateAndAddTxToPool(tx)
-		// We ignore errors here, since they mean that the tx is no longer valid and we shouldn't re-add it
-		if err == nil {
-			m.With().Info("transaction from contextually invalid block re-added to mempool",
-				log.TxId(tx.Id().ShortString()))
-		}
-	}
```
Comment (noamnelke) [inline mesh/mesh.go:387]: This logic returns transactions from contextually invalid blocks into the mempool - why remove it?

Without it, I can potentially censor transactions in busy times (where not all transactions make it into a block in the first layer) by including them in a block and sending it late. This will result in nobody ever including this transaction in a block in the future (because it can't be retransmitted, due to our old messages mechanism).

File: mesh/mesh.go [hunk 6]
```diff
@@ -282,9 +332,9 @@ func (m *Mesh) persistLayerHash() {
 	}
 }
 
-func (m *Mesh) ExtractUniqueOrderedTransactions(l *types.Layer) (validBlockTxs, invalidBlockTxs []*types.Transaction) {
-	// Separate blocks by validity
-	validBlocks, invalidBlocks := m.BlocksByValidity(l.Blocks())
+func (m *Mesh) ExtractUniqueOrderedTransactions(l *types.Layer) (validBlockTxs []*types.Transaction) {
+	// this flow is deactivated since we assume all blocks that get here either passed hare or tortoise
```
Comment (noamnelke) [inline mesh/mesh.go]: This comment is unclear if you don't know what flow was here before.

File: mesh/mesh.go [hunk 7]
```diff
@@ -302,7 +352,7 @@ func (m *Mesh) ExtractUniqueOrderedTransactions(l *types.Layer) (validBlockTxs,
 
 	// Get and return unique transactions
 	seenTxIds := map[types.TransactionId]struct{}{}
-	return m.getTxs(uniqueTxIds(validBlocks, seenTxIds), l), m.getTxs(uniqueTxIds(invalidBlocks, seenTxIds), l)
```
Comment (noamnelke) [inline mesh/mesh.go:336]: See other comments for why I think this code was not adequately migrated elsewhere.

What the removed code returns is the TxIDs that are only in invalid blocks. I don't think this logic was replaced anywhere.

File: mesh/mesh_test.go [hunk 8]
```diff
@@ -383,18 +379,15 @@ func TestMesh_ExtractUniqueOrderedTransactions(t *testing.T) {
...
-	addBlockWithTxs(r, msh, layerID, false, tx5)
+	addBlockWithTxs(r, msh, layerID, false, tx4)
 	l, err := msh.GetLayer(layerID)
 	r.NoError(err)
 
-	validBlocks, invalidBlocks := msh.ExtractUniqueOrderedTransactions(l)
+	validBlocks := msh.ExtractUniqueOrderedTransactions(l)
 
 	r.ElementsMatch(GetTransactionIds(tx1, tx2, tx3, tx4), GetTransactionIds(validBlocks...))
-	r.ElementsMatch(GetTransactionIds(tx5), GetTransactionIds(invalidBlocks...))
```
Comment (noamnelke) [inline mesh/mesh_test.go:397]: This is some pretty delicate logic that was seemingly lost :(

File: pending_txs/pending_txs.go [hunk 9]
```diff
@@ -43,23 +43,11 @@ func (apt *AccountPendingTxs) Add(layer types.LayerID, txs ...*types.Transaction
...
-		if found {
-			if existing[tx.Id()].HighestLayerIncludedIn > layer {
-				continue
-			}
-			delete(existing, tx.Id())
-			if len(existing) == 0 {
-				delete(apt.PendingTxs, tx.AccountNonce)
-			}
-		}
-	}
```
Comment (noamnelke) [inline pending_txs/pending_txs.go]: This logic was here for a reason - I don't understand why it was removed.

It's also not done cleanly - the property `HighestLayerIncludedIn` is only ever used in this code, so if it's removed the property and the code that updates it should also be removed.

File: pending_txs/pending_txs_test.go [hunk 10]
```diff
@@ -90,31 +90,31 @@ func TestNewAccountPendingTxs(t *testing.T) {
...
+		r.Equal(prevBalance-100, balance)
+
+		// However, rejecting a transaction in the highest layer it was seen in DOES remove it
+		pendingTxs.Remove(nil, []*types.Transaction{
+			newTx(t, 5, 100, 1),
+		}, 2)
+		nonce, balance = pendingTxs.GetProjection(prevNonce, prevBalance)
+		r.Equal(prevNonce, nonce)
+		r.Equal(prevBalance, balance)
+	*/
```
Comment (noamnelke) [inline pending_txs/pending_txs_test.go]: I think we should restore this logic, but if we end up not doing that - we should remove the commented out tests.

File: pending_txs/pending_txs_test.go [hunk 11]
```diff
@@ -145,12 +145,13 @@ func TestNewAccountPendingTxs(t *testing.T) {
...
-	r.Equal(prevBalance-50-950, balance)
+	/*
+		// Rejecting a transaction only removes that version, if several exist
+		// This can also cause a transaction that would previously over-draft the account to become valid
+		pendingTxs.Remove(nil, []*types.Transaction{
+			newTx(t, 5, 100, 2),
+		}, 2)
+		nonce, balance = pendingTxs.GetProjection(prevNonce, prevBalance)
+		r.Equal(int(prevNonce)+2, int(nonce))
+		r.Equal(prevBalance-50-950, balance)*/
```
Comment (noamnelke) [inline pending_txs/pending_txs_test.go]: same

**Review by noamnelke — CHANGES_REQUESTED** [review]

File: mesh/meshdb.go (hunk 4 above)
Comment (antonlerner) [reply mesh/meshdb.go:476]: where does this happen? perhaps I've missed this part but I think we should revise this to update pending transactions only from contextually valid or hare blocks

File: mesh/mesh.go (hunk 5 above)
Comment (antonlerner) [reply mesh/mesh.go:387]: Since there could be no contextually invalid blocks in transaction processor anymore, these transactions should not be taken from the mempool at all, making it easier to manage the tx's and not having to remove and add all the time

File: mesh/mesh.go (hunk 6 above)
Comment (antonlerner) [reply mesh/mesh.go]: as I see it, when dealing only with contextually valid blocks this makes the process a lot easier and understandable

File: pending_txs/pending_txs_test.go (hunk 10 above)
Comment (antonlerner) [reply pending_txs/pending_txs_test.go]: this is exactly why I've keps it, lets talk about this logic, see if it is indeed needed... Since I've managed to pass all tests except these regarding rejected transactions

File: mesh/mesh.go [hunk 12]
```diff
@@ -263,15 +282,37 @@ func (m *Mesh) HandleValidatedLayer(validatedLayer types.LayerID, layer []types.
 	for _, blockId := range layer {
 		block, err := m.GetBlock(blockId)
 		if err != nil {
-			//todo: can this happen?
-			log.Panic("hare terminated with block that is not present in mesh")
+			//stop processing this hare result, wait until tortoise pushes this layer into state
```
Comment (noamnelke) [inline mesh/mesh.go:317]: This is fine for now, but I think we should create an issue to better handle this case.

Specifically we should have some escalation mechanism, e.g.: (1) wait 30 seconds and try again (2) if some blocks still missing, ask your neighbors for them (3) if none of your neighbors have them, refresh your neighbors and try again (4) still no luck? do that a few times (5) panic, because our basic assumptions have been violated.

cc: @barakshani

**Comment by antonlerner** [comment]
bors try

File: mesh/reward_test.go [hunk 13]
```diff
@@ -278,22 +278,6 @@ func copyLayer(t *testing.T, srcMesh, dstMesh *Mesh, dstAtxDb *AtxDbMock, id typ
 	return blockIds
 }
 
-func Test_HandleValidatedLayer_panic(t *testing.T) {
```
Comment (noamnelke) [inline mesh/reward_test.go]: why remove this test?

**Comment by antonlerner** [comment]
bors try

File: mesh/mesh.go (hunk 2 above)
Comment (noamnelke) [reply mesh/mesh.go:270]: This is related to our discussion here: https://github.com/spacemeshos/go-spacemesh/issues/1876

I think we should create an issue to look into this.

File: mesh/mesh.go [hunk 14]
```diff
@@ -266,19 +270,106 @@ func (m *Mesh) pushLayersToState(oldPbase types.LayerID, newPbase types.LayerID)
 	for layerId := oldPbase; layerId < newPbase; layerId++ {
 		l, err := m.GetLayer(layerId)
 		if err != nil || l == nil {
-			// TODO: propagate/handle error
 			m.With().Error("failed to get layer", log.LayerId(layerId.Uint64()), log.Err(err))
-			break
+			return
```
Comment (noamnelke) [inline mesh/mesh.go:276]: Changing this from `break` to `return` means we won't persist any layers that did succeed, but since the state would be fucked up in this scenario, perhaps `panic`ing here is just as good?

File: mesh/mesh.go (hunk 6 above)
Comment (noamnelke) [reply mesh/mesh.go]: My comment was specifically about the wording "this flow is deactivated". What flow?

The previous comment said "Separate blocks by validity" so I understand that you mean "Separating blocks by validity is not required because we assume all blocks that get here either passed hare or tortoise". If I understood it correctly - maybe that's what the comment should say (because I wouldn't understand it without seeing the previous version). If I didn't interpret it correctly - then please clarify (in the comment).

**Comment by antonlerner** [comment]
bors try

File: mesh/mesh.go [hunk 15]
```diff
@@ -266,19 +270,106 @@ func (m *Mesh) pushLayersToState(oldPbase types.LayerID, newPbase types.LayerID)
...
 		}
-		m.AccumulateRewards(l, m.config)
-		m.PushTransactions(l)
-		m.logStateRoot(layerId)
-		m.setLayerHash(l)
-		m.setLatestLayerInState(layerId)
+		validBlocks, invalidBlocks := m.BlocksByValidity(l.Blocks())
+		m.reInsertTxsToPool(validBlocks, invalidBlocks, l.Index())
+		m.updateStateWithLayer(layerId, types.NewExistingLayer(layerId, validBlocks))
```
Comment (noamnelke) [inline mesh/mesh.go]: The order here must be flipped. When performing validation of transactions before putting them in the pool we look at the current global state. We must consider the recently applied transactions when making this validation.

File: mesh/mesh.go [hunk 16]
```diff
@@ -266,19 +270,106 @@ func (m *Mesh) pushLayersToState(oldPbase types.LayerID, newPbase types.LayerID)
...
+			//stop processing this hare result, wait until tortoise pushes this layer into state
+			log.Error("hare terminated with block that is not present in mesh")
+			return
+		}
+		blocks = append(blocks, block)
+	}
+	lyr := types.NewExistingLayer(validatedLayer, blocks)
+	invalidBlocks := m.getInvalidBlocksByHare(lyr)
+	m.reInsertTxsToPool(blocks, invalidBlocks, lyr.Index())
+	m.updateStateWithLayer(validatedLayer, lyr)
```
Comment (noamnelke) [inline mesh/mesh.go]: Flip order here as well.

File: mesh/mesh_test.go (hunk 8 above)
Comment (noamnelke) [reply mesh/mesh_test.go:397]: Let's recreate this test with the new logic (essentially test that `reInsertTxsToPool` does the right thing with all the edge cases that were originally in this test (all kinds of overlapping transactions between valid/invalid blocks).

**Review by noamnelke — CHANGES_REQUESTED** [review]

**Comment by antonlerner** [comment]
bors try

**Comment by antonlerner** [comment]
bors try

**Comment by antonlerner** [comment]
bors try

File: mesh/mesh_test.go [hunk 17]
```diff
@@ -363,15 +366,46 @@ func TestMesh_AddBlockWithTxs_PushTransactions_UpdateUnappliedTxs(t *testing.T)
...
+	blocks = append(blocks, addBlockWithTxs(r, msh, layerID, true, tx4, tx5))
+	blocks = append(blocks, addBlockWithTxs(r, msh, layerID, true, tx5))
+	hareBlocks := blocks[:3]
+
+	invalid := msh.getInvalidBlocksByHare(types.NewExistingLayer(layerID, hareBlocks))
+	r.ElementsMatch(blocks[3:], invalid)
+
+	msh.reInsertTxsToPool(blocks[3:], invalid, layerID)
+
+}
```
Comment (noamnelke) [inline mesh/mesh_test.go]: I think assertions are missing at this point? Otherwise we're only checking that this doesn't panic...

File: mesh/mesh_test.go (hunk 8 above)
Comment (noamnelke) [reply mesh/mesh_test.go:397]: I still don't see that this is tested. I can live with leaving it untested if you feel this would be hard with the new way this is written, because I expect this logic to change in the next couple of months, but I would sleep better if this logic was tested, so just making sure this wasn't just forgotten.

By "this logic" I mean: if a transaction appears in both a valid and an invalid block (`tx4`) - it's treated only as valid, but if a transaction appears only in an invalid block (`tx5`) it's treated only as invalid. Txs that appear in only one (`tx1`, `tx3`) or more (`tx2`) valid blocks are also treated as valid, and only handled once.

**Review by noamnelke — APPROVED** [review]
I still think the tests are less extensive after this change than before, but I leave it up to you to add more tests / assertions or merge it the way it is.

File: mesh/mesh_test.go (hunk 17 above)
Comment (antonlerner) [reply mesh/mesh_test.go]: good catch, I forgot to move this here
`r.ElementsMatch(GetTransactionIds(tx5), GetTransactionIds(invalidBlocks...))`
this is the part where this is supposed to be tested now, sinec validity of txs happens before the transaction flow

File: mesh/mesh_test.go [hunk 18]
```diff
@@ -395,14 +395,15 @@ func TestMesh_AddBlockWithTxs_PushTransactions_getInvalidBlocksByHare(t *testing
 	var blocks []*types.Block
 	blocks = append(blocks, addBlockWithTxs(r, msh, layerID, true, tx1, tx2))
 	blocks = append(blocks, addBlockWithTxs(r, msh, layerID, true, tx2, tx3, tx4))
-	blocks = append(blocks, addBlockWithTxs(r, msh, layerID, true, tx4, tx5))
+	blocks = append(blocks, addBlockWithTxs(r, msh, layerID, true, tx4))
 	blocks = append(blocks, addBlockWithTxs(r, msh, layerID, true, tx5))
 	hareBlocks := blocks[:3]
 
 	invalid := msh.getInvalidBlocksByHare(types.NewExistingLayer(layerID, hareBlocks))
 	r.ElementsMatch(blocks[3:], invalid)
```
Comment (noamnelke) [inline mesh/mesh_test.go]: The original test also checked that a transaction that appear in both valid and invalid blocks are treated as valid. I don't think it's super critical, but fixing this is easy (in the original test there were two valid and one invalid blocks):
```suggestion
	blocks = append(blocks, addBlockWithTxs(r, msh, layerID, true, tx4, tx5))
	hareBlocks := blocks[:2]

	invalid := msh.getInvalidBlocksByHare(types.NewExistingLayer(layerID, hareBlocks))
	r.ElementsMatch(blocks[2:], invalid)
```

(I've run this modification locally to ensure it works, so it should be safe to commit my suggestion)

**Review by noamnelke — APPROVED** [review]
I'm ok with merging it with/without my latest suggestion--up to you.

**Comment by antonlerner** [comment]
bors merge

**Comment by antonlerner** [comment]
bors merge
